### PR TITLE
[2.0] Add cpp-package

### PR DESCRIPTION
**Pull request by barry-jin**
## Description ##
Migrate cpp-package to MXNet2.0. Still work in progress. 

### Essentials ###
- [x] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

### Changes ###
- [x] Adopt CachedOp API on graph executor
- [x] Adopt autograd API on backward pass
- [x] Port SoftmaxOutput operator to 2.0
- [x] Test in CI
- [ ] Xavier initailizer will result in gradient vanish, switch to use Uniform and need more investigation. 
- [x] Add Documentation
- [ ] Refactor OpWrapperGenerator.py to solve cross-compilation problems

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by barry-jin** [comment]
@mxnet-bot run ci [sanity, unix-gpu]

**Comment by barry-jin** [comment]
@mxnet-bot run ci [sanity, unix-gpu]

**Comment by barry-jin** [comment]
@mxnet-bot run ci [unix-gpu]

File: cpp-package/scripts/OpWrapperGenerator.py [hunk 1]
```diff
@@ -0,0 +1,499 @@
...
+    MXNET_DLL int MXSymbolGetAtomicSymbolInfo(AtomicSymbolCreator creator,
+                                              const char **name,
+                                              const char **description,
+                                              mx_uint *num_args,
+                                              const char ***arg_names,
+                                              const char ***arg_type_infos,
+                                              const char ***arg_descriptions,
+                                              const char **key_var_num_args);
+    """
+    cdll.libmxnet = cdll.LoadLibrary(find_lib_path()[0])
```
Comment (leezu) [inline cpp-package/scripts/OpWrapperGenerator.py:323]: Basing the cpp-package on the ability to `dlopen` libmxnet at compile-time makes cross-compiling impossible unless the OS is setup for emulating foreign architecture code (ie. transparently running aarch64 on x86). That's not very common. Can we design the cpp package so that it integrates well with the normal compilation workflows?

**Comment by barry-jin** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by barry-jin** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by barry-jin** [comment]
@mxnet-bot run ci [centos-gpu, miscellaneous, unix-cpu, unix-gpu]

File: ci/docker/runtime_functions.sh [hunk 2]
```diff
@@ -320,6 +320,8 @@ build_ubuntu_cpu_openblas() {
         -DUSE_BLAS=Open \
         -DUSE_ONEDNN=OFF \
         -DUSE_CUDA=OFF \
+        -DUSE_CPP_PACKAGE=ON \
+        -DMXNET_USE_CPU=ON \
```
Comment (szha) [inline ci/docker/runtime_functions.sh:324]: I'm not sure whether this build option makes much sense as mxnet always has cpu mode enabled. For the lenet example, it would be great to have the code detect the GPU count instead

**Comment by barry-jin** [comment]
C++ package for master branch is now ready for review. 
Changes are listed as follows: 

1. Adopt `MXCreateCachedOp` in Executor constructor in `cpp-package/include/mxnet-cpp/executor.hpp`
2. Adopt `MXInvokeCachedOp`, `MXAutogradSetIsTraining` and `MXAutogradSetIsRecording` APIs on executor forward pass to execute graph. 
3. Adopt `MXAutogradBackwardEx` API on executor backward pass. 
4. Add tests back to CI(unix-gpu) to run some examples in `cpp-package/tests/ci_test.sh`
5. Add documentation back and add instructions for cross-compilation for cpp-package. 
Screen Shot 2021-05-05 at 2 54 18 PM

**Comment by szha** [comment]
@mxnet-bot run ci [centos-gpu, windows-cpu, unix-gpu]

File: cpp-package/README.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+### Steps to build the C++ package:
+1.  Building the MXNet C++ package requires building MXNet from source.
+2.  Clone the MXNet GitHub repository **recursively** to ensure the code in submodules is available for building MXNet.
+	```
+	git clone --recursive https://github.com/apache/incubator-mxnet mxnet
+	```
+
+3.  Install the [prerequisites](<https://mxnet.apache.org/install/build_from_source#prerequisites>), desired [BLAS libraries](<https://mxnet.apache.org/install/build_from_source#blas-library>) and optional [OpenCV, CUDA, and cuDNN](<https://mxnet.apache.org/install/build_from_source#optional>) for building MXNet from source.
+4.  There is a configuration file for make, [make/config.mk](<https://github.com/apache/incubator-mxnet/blob/master/make/config.mk>) that contains all the compilation options. You can edit this file and set the appropriate options prior to running the **make** command.
```
Comment (szha) [inline cpp-package/README.md:36]: the build instructions are no longer accurate.

File: cpp-package/example/README.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+## Building C++ examples
+
+The examples in this folder demonstrate the **training** workflow. The **inference workflow** related examples can be found in [inference](<https://github.com/apache/incubator-mxnet/blob/master/cpp-package/example/inference>) folder.
+Please build the MXNet C++ Package as explained in the [README](<https://github.com/apache/incubator-mxnet/tree/master/cpp-package#building-c-package>) File before building these examples manually.
+The examples in this folder are built while building the MXNet library and cpp-package from source. However, they can be built manually as follows
+
+From cpp-package/examples directory
+
+-  Build all examples in release mode: **make all**
+-  Build all examples in debug mode: **make debug**
```
Comment (szha) [inline cpp-package/example/README.md:29]: switch to cmake

File: cpp-package/include/mxnet-cpp/CPPLINT.cfg [hunk 5]
```diff
@@ -0,0 +1,19 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Comment (szha) [inline cpp-package/include/mxnet-cpp/CPPLINT.cfg:1]: we had cpplint in sanity check before so it's probably a good idea to bring it back.

File: cpp-package/tests/ci_test.sh [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+./test_score 0.93
+
+cp /work/build/cpp-package/example/test_ndarray_copy .
+./test_ndarray_copy
+
+# skippping temporarily, tracked by https://github.com/apache/incubator-mxnet/issues/20011
+#cp ../../build/cpp-package/example/test_regress_label .
+#./test_regress_label
+
+# sh unittests/unit_test_mlp_csv.sh
```
Comment (leezu) [inline cpp-package/tests/ci_test.sh:67]: Have you tried running?

File: cpp-package/tests/travis/run_test.sh [hunk 7]
```diff
@@ -0,0 +1,42 @@
+#!/bin/bash
```
Comment (leezu) [inline cpp-package/tests/travis/run_test.sh:1]: Please don't add unused files

File: src/operator/softmax_output-inl.h [hunk 8]
```diff
@@ -0,0 +1,476 @@
+/*
```
Comment (leezu) [inline src/operator/softmax_output-inl.h:1]: These files were removed in https://github.com/apache/incubator-mxnet/pull/18531
What's the intention of adding them back? If you like to reconsider the plan for supporting the Module API in MXNet 2, why just add back SoftmaxOutput and not the other loss function? Should this be discussed in the https://github.com/apache/incubator-mxnet/issues/16167 RFC (ie. providing an update that you want to support the Module API?)

File: cpp-package/example/inference/unit_test_imagenet_inference.sh [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    echo ">>> INFO: FP32 dummy data"
+    DYLD_LIBRARY_PATH=${DYLD_LIBRARY_PATH}:../../../lib ./imagenet_inference --symbol_file "./model/Inception-BN-symbol.json" --batch_size 1 --num_inference_batches 500 --benchmark
+else
+    echo ">>> INFO: FP32 real data"
+    LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:../../../lib ./imagenet_inference --symbol_file "./model/Inception-BN-symbol.json" --params_file "./model/Inception-BN-0126.params" --dataset "./data/val_256_q90.rec" --rgb_mean "123.68 116.779 103.939" --batch_size 1 --num_skipped_batches 50 --num_inference_batches 500
+
+    echo ">>> INFO: FP32 dummy data"
+    LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:../../../lib ./imagenet_inference --symbol_file "./model/Inception-BN-symbol.json" --batch_size 1 --num_inference_batches 500 --benchmark
+
+    lib_name=$(ls -a ../../../lib | grep -oE 'mkldnn' | tail -1)
```
Comment (leezu) [inline cpp-package/example/inference/unit_test_imagenet_inference.sh:56]: You need to check this. Likely should be onednn

File: cpp-package/include/mxnet-cpp/.gitignore [hunk 10]
```diff
@@ -0,0 +1,2 @@
+# Rebuildable file(s)
+op.h
```
Comment (leezu) [inline cpp-package/include/mxnet-cpp/.gitignore:2]: Why is op.h placed inside the build dir? Usually we only support out of source build in cmake

File: cpp-package/README.md [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+# MXNet - C++ API
+
+The MXNet C++ Package provides C++ API bindings to the users of MXNet.  Currently, these bindings are not available as standalone package.
+The users of these bindings are required to build this package as mentioned below.
+
+## Building C++ Package
+
+The cpp-package directory contains the implementation of C++ API. As mentioned above, users are required to build this directory or package before using it.
+**The cpp-package is built while building the MXNet shared library, *libmxnet.so*.**
```
Comment (leezu) [inline cpp-package/README.md:26]: There is no reason for that. cpp-package depends only on libmxnet.so C APIs. It would be better to keep a separate CMakeLists.txt for the cpp-package with the only requirement to find libmxnet.so. This implies removing the `USE_CPP_PACKAGE` in the main CMakeLists.txt

File: ci/docker/runtime_functions.sh [hunk 12]
```diff
@@ -625,12 +625,16 @@ build_ubuntu_gpu_onednn_nocudnn() {
 build_ubuntu_gpu() {
     set -ex
     cd /work/build
+    ln -s -f /usr/local/cuda/targets/x86_64-linux/lib/stubs/libcuda.so libcuda.so.1
+    export LIBRARY_PATH=${LIBRARY_PATH}:/work/build
+    export LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:/work/build
```
Comment (leezu) [inline ci/docker/runtime_functions.sh:630]: This should be removed after https://github.com/apache/incubator-mxnet/issues/17858
Would you add a comment pointing out the temporary nature of this?

**Review by leezu — CHANGES_REQUESTED** [review]
You're adding back a subset of the operators removed in https://github.com/apache/incubator-mxnet/pull/18531 ("These operators are often used with mx.module APIs. Removing them for mxnet 2.0. You can find equivalent loss functions in mx.gluon.loss namespace.") Why not add back all the Operators that are commonly used therein? This may also be worth an update in the MXNet 2 RFC https://github.com/apache/incubator-mxnet/issues/16167 as you propose to support the Symbol APIs in the CPP package. For example, this change in CPP package also means that there is no need to drop the other language bindings that were also removed in https://github.com/apache/incubator-mxnet/pull/18531

File: cpp-package/README.md (hunk 11 above)
Comment (barry-jin) [reply cpp-package/README.md:26]: Thanks for pointing it out. Let me update the documentation.

File: cpp-package/include/mxnet-cpp/.gitignore (hunk 10 above)
Comment (barry-jin) [reply cpp-package/include/mxnet-cpp/.gitignore:2]: op.h is generated by OpWrapperGenerator.py from user side, which should not be tracked by git. Should I just put this .gitignore in cpp-package directory or merged with incubator-mxnet/.gitignore

Comment (leezu) [reply cpp-package/include/mxnet-cpp/.gitignore:2]: > op.h is generated by OpWrapperGenerator.py from user side

OpWrapperGenerator.py is invoked by cmake inside the `build` folder ("out of source build"). Usually out-of-source build should not modify the code / files outside of the `build` folder.

**Comment by barry-jin** [comment]
@mxnet-bot run ci [centos-gpu, unix-gpu]

**Review by leezu — APPROVED** [review]
Thank you for adding the missing operators. I think the PR is good to merge now and the other points can be addressed later
